pkg/apis/experimental/types.go

```diff
@@ -508,35 +511,70 @@ type IngressStatus struct {
 	LoadBalancer api.LoadBalancerStatus `json:"loadBalancer,omitempty"`
 }
 
-// IngressRule represents the rules mapping the paths under a specified host to the related backend services.
-type IngressRule struct {
-	// Host is the fully qualified domain name of a network host, or its IP
-	// address as a set of four decimal digit groups separated by ".".
-	// Conforms to RFC 1738.
+// IngressHost represents the rules mapping the paths under a specified host to
+// the related backend services.
+type IngressHost struct {
+	// Host is the fully qualified domain name of a network host, as defined
+	// by RFC 3986. Note the following deviations from the "host" part of the
+	// URI as defined in the RFC:
+	// 1. IPs are not allowed. Currently an IngressRule can only apply to the
+	//	  IP in the Spec of the parent Ingress.
+	// 2. The `:` delimiter is not respected because ports are not allowed.
+	//	  Currently the port of an Ingress is implicitly :80 for http and
+	//	  :443 for https.
+	// Both these may change in the future.
+	// Incoming requests are matched against the Host before the IngressRule.
 	Host string `json:"host,omitempty"`
+	// IngressRule represents a rule to route requests for this IngressHost.
+	IngressRule `json:",inline"`
+}
+
+// IngressRule represents a rule to apply against incoming requests. If the
+// rule is satisfied, the request is routed to the specified backend.
+type IngressRule struct {
+	//TODO:
+	// 1. Consider renaming this resource and the associated rules so they
+	// aren't tied to Ingress. They can be used to route intra-cluster traffic.
+	// 2. Consider adding fields for ingress-type specific global options
+	// usable by a loadbalancer, like http keep-alive.
 
-	// Paths describe a list of load-balancer rules under the specified host.
-	Paths []IngressPath `json:"paths"`
+	// Currently mixing different types of rules in a single Ingress is
+	// disallowed, so exactly one of the following must be set.
+	HTTP *HTTPIngressRule `json:"http"`
+}
+
+// HTTPIngressRule is a list of http selectors pointing to IngressBackends.
+// In the example: http://<host>/<path>?<searchpart> -> IngressBackend where
+// where parts of the url correspond to RFC 3986, this resource will be used
+// to match against everything after the last '/'.
```

Comment:
That's lb controller specific. We don't validate against it (path should be a valid regex, that's all -- see comment above validation where I call CompilePOSIX). If the other end validates it as a pure path it should surface the error. Otherwise you can totally specify a full url regex in haproxy with the right config options.